heroic-core/src/main/java/com/spotify/heroic/metric/LocalMetricManager.java

```diff
@@ -569,10 +577,11 @@ public FullQuery end(int resolved, int failed, int cancelled) throws Exception {
     }
 
     @RequiredArgsConstructor
-    private static class QuotaWatcher implements FetchQuotaWatcher, RetainQuotaWatcher {
+    private static class ReadWatcher implements FetchQuotaWatcher, RetainQuotaWatcher {
         private final long dataLimit;
         private final long retainLimit;
         private final DataInMemoryReporter dataInMemoryReporter;
+        private final ConcurrentHashMultiset<Long> rowDensityData = ConcurrentHashMultiset.create();
```

Comment:
Hate to be a stickler, but this should not live in the watcher anymore since it's not being reported by the backend :). A better place would be in the `ResultCollector`.